library/std/src/thread/scoped.rs

```diff
@@ -51,12 +54,24 @@ impl ScopeData {
             panic!("too many running threads in thread scope");
         }
     }
+    fn main_thread(&self) -> &Thread {
+        unsafe { &*self.main_thread.get() }
+    }
     pub(super) fn decrement_num_running_threads(&self, panic: bool) {
         if panic {
             self.a_thread_panicked.store(true, Ordering::Relaxed);
         }
+        let main_thread = self.main_thread().clone();
         if self.num_running_threads.fetch_sub(1, Ordering::Release) == 1 {
-            self.main_thread.unpark();
+            // By now, `num_running_threads` is `0`, so when `scope()` in the main thread
+            // is unparked, it will complete its business and deallocate `*self`!
+            // Two things to look after:
+            //   - it can spuriously unpark / wake up, **so `self` can no longer be used**, even
+            //     before we, ourselves, unpark. Hence why we've cloned the `main_thread`'s handle.
+            //   - no matter how it unparks, `*self` may be deallocated before this function
+            //     returns, so all of `*self` data, **including `main_thread`**, must be interiorly
+            //     mutable. See https://github.com/rust-lang/rust/issues/55005 for more info.
```

Comment:
I'd suggest to link to https://github.com/rust-lang/rust/pull/98017, which adjusts the UnsafeCell docs for this.